### PR TITLE
issue #41 provide API for Rules #41

### PR DESCRIPTION
**Pull request by jojowork**
It is possible to import/export Events of type template via APIv4.
But the EventMessage-Rules are not available via API.
To be able to import/export EventMessage-Rules an APIv4 Endpoint is necessary for them.

this PR is supposed to enhance eventmessages extension with an API for Event Message Rules

### CODE REVIEW

**Comment by jensschuppe** [comment]
Nice work so far! As discussed, this needs to make sure that the schema matches the current DB table structure, or adjusts it with an upgrade task.

**Comment by jojowork** [comment]
I designed the Schema-XML: [xml/schema/CRM/Eventmessages/EventMessageRule.xml](https://github.com/systopia/de.systopia.eventmessages/pull/42/commits/eaba2aa0f8b2ffefd1512ef5ce2f9f441f3e89b5#diff-8d2ff9e43152364bcb7e05f34853c3ea4046331de9cc4689bbbce1d9d3d49cba)
to match with the SQL-Schema definition: https://github.com/systopia/de.systopia.eventmessages/blob/master/sql/civicrm_value_event_messages.sql

I think that it matches. Or do you see any deviations?

**Comment by jensschuppe** [comment]
I haven't reviewed it yet, just wanted to note that for reviewers. In any case we'll want to use the auto-generated `auto_install.sql` in favor of the old SQL file and review for code references.

File: CRM/Eventmessages/BAO/EventMessageRule.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    $hook = empty($params['id']) ? 'create' : 'edit';
+
+    CRM_Utils_Hook::pre($hook, $entityName, CRM_Utils_Array::value('id', $params), $params);
+    $instance = new $className();
+    $instance->copyValues($params);
+    $instance->save();
+    CRM_Utils_Hook::post($hook, $entityName, $instance->id, $instance);
+
+    return $instance;
+  } */
```
Comment (dontub) [inline CRM/Eventmessages/BAO/EventMessageRule.php:24]: I'd suggest to remove the commented lines.

File: eventmessages.civix.php [hunk 2]
```diff
@@ -24,7 +24,7 @@ class CRM_Eventmessages_ExtensionUtil {
    *   Translated text.
    * @see ts
    */
-  public static function ts($text, $params = []): string {
+  public static function ts($text, $params = []) {
```
Comment (dontub) [inline eventmessages.civix.php:27]: This change is not necessary. (Are you using an older version of civix?)

File: eventmessages.civix.php [hunk 3]
```diff
@@ -41,7 +41,7 @@ public static function ts($text, $params = []): string {
    *   Ex: 'http://example.org/sites/default/ext/org.example.foo'.
    *   Ex: 'http://example.org/sites/default/ext/org.example.foo/css/foo.css'.
    */
-  public static function url($file = NULL): string {
+  public static function url($file = NULL) {
```
Comment (dontub) [inline eventmessages.civix.php:44]: Not necessary. Seems that the same holds true for the other changes in this file. If you think this file should be updated, do this in a separate PR, please.

File: xml/schema/CRM/Eventmessages/EventMessageRule.xml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    <name>event_id</name>
+    <table>civicrm_event</table>
+    <key>id</key>
+    <onDelete>CASCADE</onDelete>
+  </foreignKey>
+
+  <field>
+    <name>is_active</name>
+    <type>int unsigned</type>
+    <comment>FK to Event</comment>
```
Comment (dontub) [inline xml/schema/CRM/Eventmessages/EventMessageRule.xml:40]: Should be the same as in the [current schema](https://github.com/systopia/de.systopia.eventmessages/blob/b057912f8db70be3cdeef3361cd9933a87cb5f43/sql/civicrm_value_event_messages.sql#L5C73-L5C73): `is this rule active`.

File: xml/schema/CRM/Eventmessages/EventMessageRule.xml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  </field>
+
+  <field>
+    <name>template_id</name>
+    <type>int unsigned</type>
```
Comment (dontub) [inline xml/schema/CRM/Eventmessages/EventMessageRule.xml:45]: This field should be required.

**Review by dontub — CHANGES_REQUESTED** [review]
Thanks for the PR!
Can you additionally drop [this line](https://github.com/systopia/de.systopia.eventmessages/blob/b057912f8db70be3cdeef3361cd9933a87cb5f43/CRM/Eventmessages/Upgrader.php#L29) as it's now done via `auto_install.sql`?

**Comment by jojowork** [comment]
I made the mentioned changes.

**Comment by jojowork** [comment]
Achtung: Hier muss noch getestet werden, ob bereits erstellte Rules auch nach den Änderungen erhalten bleiben.

**Comment by dontub** [comment]
@jojowork, you forgot to run `civix generate:entity-boilerplate` for your change in https://github.com/systopia/de.systopia.eventmessages/pull/42/commits/f9bfbcbb9dbde084e13ca8ea99ef6d317387f3ee.

**Comment by jojowork** [comment]
> @jojowork, you forgot to run `civix generate:entity-boilerplate` for your change in [f9bfbcb](https://github.com/systopia/de.systopia.eventmessages/commit/f9bfbcbb9dbde084e13ca8ea99ef6d317387f3ee).

I updated the EventMessageRule.xml to match as good as possible to the old https://github.com/systopia/de.systopia.eventmessages/blob/master/sql/civicrm_value_event_messages.sql and then executed `civix generate:entity-boilerplate`

I hope everything is fine now.

**Pull request merged**